### PR TITLE
HHH-12338 - Incorrect metamodel for basic collections

### PR DESCRIPTION
**Pull request by vladmihalcea**
This fix is needed so that we can map List, Set, or Map entity attributes to JSON columns or to map List entity attributes to ARRAY columns, as explained [in this Hibernate Types project issue](https://github.com/vladmihalcea/hibernate-types/issues/137).

Initially, this was integrated into 5.2.17, but it was causing the [HHH-12581](https://hibernate.atlassian.net/browse/HHH-12581) issue, so it got reverted completely.

I this it's worth addressing it via a fix instead of removing this functionality altogether.

@sebersole @Naros If you have time to review it, I'd really appreciate it.

### CODE REVIEW

**Comment by yangbinfx** [comment]
unsubscribe


------------------&nbsp;原始邮件&nbsp;------------------
发件人:&nbsp;"Vlad Mihalcea"<notifications@github.com&gt;;
发送时间:&nbsp;2020年2月21日(星期五) 下午4:14
收件人:&nbsp;"hibernate/hibernate-orm"<hibernate-orm@noreply.github.com&gt;;
抄送:&nbsp;"Subscribed"<subscribed@noreply.github.com&gt;;
主题:&nbsp;[hibernate/hibernate-orm] HHH-12338 - Incorrect metamodel for basic collections (#3254)




This fix is needed so that we can map List, Set, or Map entity attributes to JSON columns or to map List entity attributes to ARRAY columns, as explained in this Hibernate Types project issue.
 
Initially, this was integrated into 5.2.17, but it was causing the HHH-12581 issue, so it got reverted completely.
 
I this it's worth addressing it via a fix instead of removing this functionality altogether.
 
@sebersole @Naros If you have time to review it, I'd really appreciate it.
 
 
You can view, comment on, or merge this pull request online at:
 
&nbsp; https://github.com/hibernate/hibernate-orm/pull/3254
 
Commit Summary
    
HHH-12338 - Incorrect metamodel for basic collections
  
File Changes
    
     M     tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java     (16)   
   
     A     tooling/metamodel-generator/src/test/java/org/hibernate/jpamodelgen/test/collectionbasictype/CollectionAsBasicTypeTest.java     (89)   
  
Patch Links:
    
https://github.com/hibernate/hibernate-orm/pull/3254.patch
   
https://github.com/hibernate/hibernate-orm/pull/3254.diff

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java
Comment (NathanQingyangXu) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java]: One-shot usage of `TypeUtils.containsAnnotation(Element, String)` is okay, but multiple invocations won't be as efficient as
```
AnnotationMirror annotationMirrors =  element.getAnnotationMirrors();
// go about checking based on `AnnotationMirror` using `public static boolean isAnnotationMirrorOfType(AnnotationMirror annotationMirror, String fqcn)` method
```

Furthermore, I do think we can reformat the above code in some way to improve readability. For instance, I spent quite some time figuring out how the parenthesis pairs match up above.

Comment (vladmihalcea) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java]: Sure thing. Feel free to modify it any way you want and push the commit.

Comment (NathanQingyangXu) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java]: I don't think I have such rights, but I can create a PR after yours is merged.

Comment (vladmihalcea) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java]: I think you can [fork my repository](https://github.com/vladmihalcea/hibernate-orm/tree/HHH-12338), and your change and send a Pull Request.

Comment (NathanQingyangXu) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java]: Thanks for the tip, but I would approve the PR if you made the code cleaning as I proposed, :).

Comment (vladmihalcea) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java]: I don't know exactly what changes you want to make, so it's better if you cherry-pick my commit, add your changes and submit a new PR to the main fork.

I'll close my PR once you do that.

Comment (NathanQingyangXu) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java]: I think 
```
if (( TypeUtils.containsAnnotation( element, Constants.CONVERT ) ||
				TypeUtils.containsAnnotation( element, Constants.HIBERNATE_TYPE ) ) &&
				!isCollectionAssociation( element )
		) {
```
could be simplified dramatically by
```
if ( TypeUtils.containsAnnotation( element, Constants.CONVERT, Constants.HIBERNATE_TYPE ) && ! TypeUtils.containsAnnotation( element, Constants.ONE_TO_MANY, Constants.MANY_TO_MANY, Constants.ELEMENT_COLLECTION) ) {

```

Comment (vladmihalcea) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java]: @NathanQingyangXu I pushed a new commit based on your changes.

Comment (NathanQingyangXu) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java]: thanks. I approved. I would let other reviewer focus on business logic stuff.

**Comment by vladmihalcea** [comment]
@NathanQingyangXu Thanks for approving the PR.

I think it's a good idea if @Naros investigate it too, especially since we now have more context about why we need this, as opposed to the other time when this issue got reverted due to a problem that's fixed now.

**Comment by chingyi-lin** [comment]
Hi! Our team is using hibernate-orm and have been seeing this error. Is there a plan to merge the fix?

**Comment by nicobao** [comment]
Hi, our team is facing the issue mentioned here: https://github.com/vladmihalcea/hibernate-types/issues/137 
I don't use vlad's library but it happens while trying to deserialize a String of JSON format from DB to a Map<String,Set<String>>.
Any chance this can be merged to master?

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java
Comment (beikov) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java:156]: I don't think this is according to the JPA specification, but I didn't look into it to find a reference. My gut tells me, this should be fixed by annotating the target attribute with `@Basic` so IMO it should look like the following:

```suggestion
		if ( TypeUtils.containsAnnotation(
				element,
				Constants.CONVERT,
				Constants.HIBERNATE_TYPE,
				Constants.BASIC
		) ) {
			return new AnnotationMetaSingleAttribute(
					entity,
					element,
					TypeUtils.toTypeString( declaredType )
			);
		}
```

**Review by beikov — CHANGES_REQUESTED** [review]

**Comment by dk98126** [comment]
Any changes in 2021? This problem still reproduces in Hibernate 5.4.27.Final

**Comment by beikov** [comment]
Thanks for the PR. I cleaned it up and merged it!

**Comment by chingyi-lin** [comment]
Thanks @beikov! Will it be planned into the next release (5.4.30)?

**Comment by lukwasatwork** [comment]
in what version that fix will be included?
I checked 5.4.30.Final and problem still exists there

**Comment by marcelstoer** [comment]
@beikov it obviously landed on master but [not (yet) on the 5.4 branch](https://github.com/hibernate/hibernate-orm/blob/5.4/tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/MetaAttributeGenerationVisitor.java#L141). What is required for that to happen?

**Comment by apb12** [comment]
thanks for fixing it

**Pull request merged**
